### ISSUE TITLE
False positive govet shadow

### ISSUE THREAD

**Issue by irbgeo** [golangci/golangci-lint#3634]
### Welcome

- [X] Yes, I'm using a binary release within 2 latest major releases. Only such installations are supported.
- [X] Yes, I've searched similar issues on GitHub and didn't find any.
- [X] Yes, I've included all information below (version, config, etc.).
- [X] Yes, I've tried with the standalone linter if available (e.g., gocritic, go vet, etc.). (https://golangci-lint.run/usage/linters/)

### Description of the problem

False positive govet in the code below

```golang
	file, err := s.download(instruction.Src)
	if err != nil {
		return err
	}

	if instruction.CheckSum != nil {
		fileCheckSum, err := s.download(instruction.CheckSum.Src) // shadow: declaration of "err" shadows declaration at line 1
		if err != nil {
			return err
		}

		if check(file, fileCheckSum, instruction.CheckSum.Type) {
			return fmt.Errorf("the file was downloaded incorrectly")
		}
	}
```

I received message by shadow on line
```
		fileCheckSum, err := s.download(instruction.CheckSum.Src) // shadow: declaration of "err" shadows declaration at line  1
```
but it is correct structure of code
 

### Version of golangci-lint

<details>

```console
$ golangci-lint --version
golangci-lint has version v1.51.2 built from (unknown, mod sum: "h1:yIcsT1X9ZYHdSpeWXRT1ORC/FPGSqDHbHsu9uk4FK7M=") on (unknown)
```

</details>

### Configuration file

<details>

```console
$ cat .golangci.yml
# options for analysis running
run:
  # default concurrency is a available CPU number
  concurrency: 10

  # timeout for analysis, e.g. 30s, 5m, default is 1m
  deadline: 2m

  # exit code when at least one issue was found, default is 1
  issues-exit-code: 1

  # include test files or not, default is true
  tests: true

  # which files to skip: they will be analyzed, but issues from them
  # won't be reported. Default value is empty list, but there is
  # no need to include all autogenerated files, we confidently recognize
  # autogenerated files. If it's not please let us know.
  skip-files:
    - easyjson
    - ".*\\.my\\.go$"
    - "../../../go/pkg/mod/"

issues:
  # Independently from option `exclude` we use default exclude patterns,
  # it can be disabled by this option. To list all
  # excluded by default patterns execute `golangci-lint run --help`.
  # Default value for this option is true.
  issues:
    exclude:
      - "Error return value of .((os\\.)?std(out|err)\\..*|.*Close|.*Flush|os\\.Remove(All)?|.*printf?|os\\.(Un)?Setenv). is not checked"
      - "exported (type|method|function) (.+) should have comment or be unexported"
      - "ST1000: at least one file in a package should have a package comment"

# output configuration options
output:
  # colored-line-number|line-number|json|tab|checkstyle, default is "colored-line-number"
  format: colored-line-number

  # print lines of code with issue, default is true
  print-issued-lines: true

  # print linter name in the end of issue text, default is true
  print-linter-name: true

# all available settings of specific linters
linters-settings:
  govet:
    # report about shadowed variables
    check-shadowing: true
  gofmt:
    # simplify code: gofmt with `-s` option, true by default
    simplify: true
  errcheck:
    # report about not checking of errors in type assetions: `a := b.(MyStruct)`;
    # default is false: such cases aren't reported by default.
    check-type-assertions: true
    # report about assignment of errors to blank identifier: `num, _ := strconv.Atoi(numStr)`;
    # default is false: such cases aren't reported by default.
    check-blank: true
  gocyclo:
    # minimal code complexity to report, 30 by default (but we recommend 10-20)
    min-complexity: 25
  misspell:
    # Correct spellings using locale preferences for US or UK.
    # Default is to use a neutral variety of English.
    # Setting locale to US will correct the British spelling of 'colour' to 'color'.
    locale: US
  prealloc:
    # XXX: we don't recommend using this linter before doing performance profiling.
    # For most programs usage of prealloc will be a premature optimization.
    # Report preallocation suggestions only on simple loops that have no returns/breaks/continues/gotos in them.
    # True by default.
    simple: true
    range-loops: true # Report preallocation suggestions on range loops, true by default
    for-loops: true # Report preallocation suggestions on for loops, false by default
  unparam:
    # Inspect exported functions, default is false. Set to true if no external program/library imports your code.
    # XXX: if you enable this setting, unparam will report a lot of false-positives in text editors:
    # if it's called for subdir of a project it can't find external interfaces. All text editor integrations
    # with golangci-lint call it on a directory with the changed file.
    check-exported: false
  wsl:
    force-err-cuddling: true
  goimports:
    # A comma-separated list of prefixes, which, if set, checks import paths
    # with the given prefixes are grouped after 3rd-party packages.
    # Default: ""
    local-prefixes: github.com/fraima/fraimactl

linters:
  disable-all: true
  enable:
    - govet
    - gofmt
    - errcheck
    - misspell
    - gocyclo
    - ineffassign
    - goimports
    - nakedret
    - unparam
    - unused
    - prealloc
    - durationcheck
    - gosimple
    - bodyclose
    - errorlint
    - maintidx
    - nakedret
    - unconvert
  fast: false
```

</details>

### Go environment

<details>

```console
$ go version && go env
go version go1.20.1 linux/amd64
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/irbgeo/.cache/go-build"
GOENV="/home/irbgeo/.config/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/irbgeo/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/irbgeo/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.20.1"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/home/irbgeo/projects/fraima/fraima/go.mod"
GOWORK=""
CGO_CFLAGS="-O2 -g"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-O2 -g"
CGO_FFLAGS="-O2 -g"
CGO_LDFLAGS="-O2 -g"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -Wl,--no-gc-sections -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build3831863189=/tmp/go-build -gno-record-gcc-switches"
```

</details>

### Verbose output of running

<details>

```console
$ golangci-lint cache clean
$ golangci-lint run -v
INFO [config_reader] Config search paths: [./ /home/irbgeo/projects/fraima/fraima /home/irbgeo/projects/fraima /home/irbgeo/projects /home/irbgeo /home /] 
INFO [config_reader] Used config file .golangci.yml 
INFO [lintersdb] Active 17 linters: [bodyclose durationcheck errcheck errorlint gocyclo gofmt goimports gosimple govet ineffassign maintidx misspell nakedret prealloc unconvert unparam unused] 
INFO [loader] Go packages loading at mode 575 (types_sizes|exports_file|files|imports|name|compiled_files|deps) took 176.210284ms 
INFO [runner/filename_unadjuster] Pre-built 0 adjustments in 747.286µs 
INFO [linters_context/goanalysis] analyzers took 12.704088258s with top 10 stages: buildir: 10.572902234s, inspect: 757.650468ms, ctrlflow: 412.242564ms, printf: 358.532856ms, buildssa: 214.849734ms, fact_purity: 161.764751ms, unconvert: 53.684353ms, unparam: 30.380576ms, goimports: 20.230312ms, S1001: 16.710214ms 
INFO [runner] Issues before processing: 7, after processing: 2 
INFO [runner] Processors filtering stat (out/in): filename_unadjuster: 7/7, severity-rules: 2/2, path_prefixer: 2/2, skip_dirs: 7/7, identifier_marker: 3/3, exclude: 3/3, nolint: 2/2, uniq_by_line: 2/2, max_same_issues: 2/2, max_from_linter: 2/2, source_code: 2/2, path_prettifier: 7/7, max_per_file_from_linter: 2/2, path_shortener: 2/2, sort_results: 2/2, cgo: 7/7, skip_files: 7/7, autogenerated_exclude: 3/7, exclude-rules: 2/3, diff: 2/2 
INFO [runner] processing took 1.561325ms with stages: nolint: 529.217µs, exclude-rules: 249.777µs, identifier_marker: 212.77µs, path_prettifier: 177.852µs, autogenerated_exclude: 164.249µs, skip_dirs: 76.463µs, source_code: 63.243µs, skip_files: 60.216µs, max_same_issues: 5.359µs, uniq_by_line: 4.757µs, cgo: 4.105µs, path_shortener: 3.765µs, filename_unadjuster: 2.522µs, max_from_linter: 2.301µs, max_per_file_from_linter: 1.878µs, severity-rules: 727ns, diff: 636ns, exclude: 585ns, sort_results: 565ns, path_prefixer: 338ns 
INFO [runner] linters took 4.350216874s with stages: goanalysis_metalinter: 4.34843815s 
internal/downloader/downloading.go:31:17: shadow: declaration of "err" shadows declaration at line 25 (govet)
                fileCheckSum, err := s.download(instruction.CheckSum.Src)
                              ^
internal/downloader/downloading.go:49:10: shadow: declaration of "err" shadows declaration at line 25 (govet)
                        data, err := os.ReadFile(filePath)
                              ^
INFO File cache stats: 22 entries of total size 23.9KiB 
INFO Memory: 47 samples, avg is 401.3MB, max is 534.7MB 
INFO Execution took 4.532548494s                  
```

</details>

### Code example or link to a public repository

<details>

```go
package downloader

import (
	"fmt"
	"io"
	"net/http"
	"os"
	"path"

	"github.com/fraima/fraimactl/internal/config"
	"github.com/fraima/fraimactl/internal/utils"
)

type downloader struct {
	client *http.Client
}

func New() *downloader {
	return &downloader{
		client: &http.Client{},
	}
}

func (s *downloader) Run(instruction config.DownloadInstruction) error {
	file, err := s.download(instruction.Src)
	if err != nil {
		return err
	}

	if instruction.CheckSum != nil {
		fileCheckSum, err := s.download(instruction.CheckSum.Src)
		if err != nil {
			return err
		}

		if check(file, fileCheckSum, instruction.CheckSum.Type) {
			return fmt.Errorf("the file was downloaded incorrectly")
		}
	}

	if instruction.Unzip.Status {
		err = unzipFile(instruction.Name, file)
		if err != nil {
			return err
		}

		for _, f := range instruction.Unzip.Files {
			filePath := getDownloadDir(instruction.Name, f)
			data, err := os.ReadFile(filePath)
			if err != nil {
				return err
			}

			err = utils.CreateFile(path.Join(instruction.HostPath, path.Base(f)), data, instruction.Permission, instruction.Owner)
			if err != nil {
				return err
			}
		}
		return nil
	}

	err = utils.CreateFile(path.Join(instruction.HostPath, instruction.Name), file, instruction.Permission, instruction.Owner)
	if err != nil {
		return err
	}
	return nil
}

func (s *downloader) download(src string) ([]byte, error) {
	resp, err := s.client.Get(src)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	return data, nil
}
```

</details>

**Comment by ldez** [comment]
Hello,

golangci-lint is a wrapper around linter, the linter rules are not handled by golangci-lint.

But I will answer anyway, in your example `err` is shadowing the declaration of `err` in `file, err := s.download(instruction.Src)` so it's not a false positive.
